src/librustc_codegen_llvm/consts.rs

```diff
@@ -451,15 +465,20 @@ impl StaticMethods for CodegenCx<'ll, 'tcx> {
             // go into custom sections of the wasm executable.
             if self.tcx.sess.opts.target_triple.triple().starts_with("wasm32") {
                 if let Some(section) = attrs.link_section {
+                    let alloc_content = match &alloc.alloc_content {
+                        // FIXME: This is not really a bug but makes it testable except on wasm
+                        None => bug!(),
```

Comment:
I'm really unsure what to do here.

Reply:
So the trivial solution would be to generate a `Vec` with the appropriate amount of memory (zero filled e.g.). This may also be the correct solution considering this is wasm, I'm not sure.

If wasm can handle an undef allocation, we need to change the `LLVMMDStringInContext` call to something not taking a pointer to an array of the correct size.

cc @alexcrichton 	you wrote the original code, do you know if wasm could handle it if we (or whether it's even possible to) create an "undef string in context"?

Reply:
I'm not sure I understand enough about this PR or what's happening to really answer the question here. Custom sections in wasm are unfortunately just very different in LLVM compared to custom sections on other targets, they're done through metadata instead of normal attributes on normal statics/etc.

I'm honestly not even sure what it means to have an uninitialized static in Rust, I thought that basically wasn't possible. I would imagine though that if that does happen it should be a compiler error or silently ignored if it's never reasonably used. Basically I would expect similar behavior across all platforms, but it's pretty niche so it doesn't really matter if it differs by a bit.

Reply:
Brief synopsis: `static A: MaybeUninit<[u8; 1 << 30]> = MaybeUninit::uninit()` creates an *initialized* static that contains *uninitialized* memory which is useful for use as an allocator for example. In that specific case, all of its memory is uninitialized and so it is not necessary to store any specific representation in the generated binary. E.g. for `elf` this will simply instruct the program loader to allocate some (zeroed, but in theory arbitrary) memory of correct size and alignment and thus the binary is only a few kB large despite its 1GB usage at runtime.

The question is basically if wasm has a similar possibility since it currently allocates a string with the representation that MIR interpretation arbitrarily contains, and places that into the binary.

Reply:
Ok thanks for the info! Wasm is unlike all other targets in that the custom sections here are not byte addressable, so they basically don't live in the same address space (and there's no native way to access them). There's no way to encode a 1GB custom section in just a few kilobytes in wasm, it has to actually be 1GB of data. That being said I don't think there's any worthwhile intention of having undefined bytes in a custom section on wasm. The "most reasonable" thing to do is to probably just allocate a bunch of zeros, but I think it would also be fine to basically just error out.

Reply:
Erroring out seems bad because it would make some completely legitimate programs fail (any where the initial value of a static contains uninitialized bytes, such as in `MaybeUninit` but also with padding).

Reply:
If I'm understanding this correctly, this situation only occurs for `link_section` statics, not normal statics (which also can have relocations), so we'd only error for those

Reply:
@oli-obk is correct and I'm saying both strategies seem reasonable, either synthesizing zeros or error'ing. If y'all thing error'ing is unacceptable then synthesizing zeros should also be fine.

Reply:
Then I'll simply fallback to a `vec![0; size]` as suggested.

Reply:
@alexcrichton I'm reading through the [specification of wasm data section](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-data) and it seems to suggest that all memory is initially guaranteed to be zeroed. You said that the `data` section can not be addresses directly, so I assume this works by the compiler setting aside some slice of memory in the main linear memory (id `0`) and inserting its initial data into the `data` section. So if one were to simply leave out the `data` entry but keep the memory range otherwise unused, then that should have the same effect as a zeroed initialization.

Reply:
@HeroicKatora you're indeed right! This is for custom sections, however, not data segments. The `#[link_section]` is interpreted as data going into custom sections for wasm (but normal `static`/`const` go into data segments). For custom sections the specification is much different than data segments.

Reply:
You need to impl the fallback